src/transport-manager.js

```diff
@@ -0,0 +1,177 @@
+const log = debug('libp2p:transports')
+log.error = debug('error:libp2p:transports')
+
+class TransportManager {
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Libp2p} options.libp2p The Libp2p instance. It will be passed to the transports.
+   * @param {Upgrader} options.upgrader The upgrader to provide to the transports
+   * @param {function(Connection)} options.onConnection Called whenever an incoming connection is received
+   */
+  constructor ({ libp2p, upgrader, onConnection }) {
+    this.libp2p = libp2p
+    this.upgrader = upgrader
+    this._transports = new Map()
+    this._listeners = new Map()
+    this.onConnection = onConnection
+  }
+
+  /**
+   * Adds a `Transport` to the manager
+   *
+   * @param {String} key
+   * @param {Transport} Transport
+   * @returns {void}
+   */
+  add (key, Transport) {
+    log('adding %s', key)
+    if (!key) {
+      throw errCode(new Error(`Transport must have a valid key, was given '${key}'`), codes.ERR_INVALID_KEY)
+    }
+    if (this._transports.has(key)) {
+      throw errCode(new Error('There is already a transport with this key'), codes.ERR_DUPLICATE_TRANSPORT)
+    }
+
+    const transport = new Transport({
+      libp2p: this.libp2p,
+      upgrader: this.upgrader
+    })
+
+    this._transports.set(key, transport)
+    this._listeners.set(key, [])
+  }
+
+  /**
+   * Stops all listeners
+   * @async
+   */
+  async close () {
+    const tasks = []
+    for (const [key, listeners] of this._listeners) {
+      log('closing listeners for %s', key)
+      while (listeners.length) {
+        tasks.push(listeners.pop().close())
+      }
+    }
+
+    await Promise.all(tasks)
+    this._listeners.clear()
+  }
+
+  /**
+   * Dials the given Multiaddr over it's supported transport
+   * @param {Multiaddr} ma
+   * @param {*} options
+   * @returns {Promise<Connection>}
+   */
+  async dial (ma, options) {
+    const transport = this.transportForMultiaddr(ma)
+    if (!transport) {
+      throw errCode(new Error(`No transport available for address ${String(ma)}`), codes.ERR_TRANSPORT_UNAVAILABLE)
+    }
+    const conn = await transport.dial(ma, options)
+    return conn
+  }
+
+  /**
+   * Returns all Multiaddr's the listeners are using
+   * @returns {Multiaddr[]}
+   */
+  getAddrs () {
+    let addrs = []
+    for (const listeners of this._listeners.values()) {
+      for (const listener of listeners) {
+        addrs = [...addrs, ...listener.getAddrs()]
+      }
+    }
+    return addrs
+  }
+
+  /**
+   * Finds a transport that matches the given Multiaddr
+   * @param {Multiaddr} ma
+   * @returns {Transport|null}
+   */
+  transportForMultiaddr (ma) {
+    for (const transport of this._transports.values()) {
+      const addrs = transport.filter([ma])
+      if (addrs.length) return transport
+    }
+    return null
+  }
+
+  /**
+   * Starts listeners for each given Multiaddr.
+   * @async
+   * @param {Multiaddr[]} addrs
+   */
+  async listen (addrs) {
```

Comment:
We can add new transports anytime during runtime. We should then be able to listen on transports added after a `listen`.

Reply:
It's totally doable in the current state, and you could also just call listen at any point if you have new addresses for existing transports.

```js
libp2p.transportManager.add(transportKey, TransportConstructor)
await libp2p.transportManager.listen(newTransportAddrs)
```